humanoid_league_team_communication/humanoid_league_team_communication/humanoid_league_team_communication.py

```diff
@@ -1,438 +1,229 @@
-                             (robocup_extension_pb2.Team.BLUE, ObstacleRelative.ROBOT_CYAN),
-                             (robocup_extension_pb2.Team.RED, ObstacleRelative.ROBOT_MAGENTA))
-        self.role_mapping = (
-            (robocup_extension_pb2.Role.ROLE_UNDEFINED, Strategy.ROLE_UNDEFINED),
-            (robocup_extension_pb2.Role.ROLE_IDLING, Strategy.ROLE_IDLING),
-            (robocup_extension_pb2.Role.ROLE_OTHER, Strategy.ROLE_OTHER),
-            (robocup_extension_pb2.Role.ROLE_STRIKER, Strategy.ROLE_STRIKER),
-            (robocup_extension_pb2.Role.ROLE_SUPPORTER, Strategy.ROLE_SUPPORTER),
-            (robocup_extension_pb2.Role.ROLE_DEFENDER, Strategy.ROLE_DEFENDER),
-            (robocup_extension_pb2.Role.ROLE_GOALIE, Strategy.ROLE_GOALIE),
-        )
-        self.action_mapping = (
-            (robocup_extension_pb2.Action.ACTION_UNDEFINED, Strategy.ACTION_UNDEFINED),
-            (robocup_extension_pb2.Action.ACTION_POSITIONING, Strategy.ACTION_POSITIONING),
-            (robocup_extension_pb2.Action.ACTION_GOING_TO_BALL, Strategy.ACTION_GOING_TO_BALL),
-            (robocup_extension_pb2.Action.ACTION_TRYING_TO_SCORE, Strategy.ACTION_TRYING_TO_SCORE),
-            (robocup_extension_pb2.Action.ACTION_WAITING, Strategy.ACTION_WAITING),
-            (robocup_extension_pb2.Action.ACTION_KICKING, Strategy.ACTION_KICKING),
-            (robocup_extension_pb2.Action.ACTION_SEARCHING, Strategy.ACTION_SEARCHING),
-            (robocup_extension_pb2.Action.ACTION_LOCALIZING, Strategy.ACTION_LOCALIZING),
-        )
-        self.side_mapping = (
-            (robocup_extension_pb2.OffensiveSide.SIDE_UNDEFINED, Strategy.SIDE_UNDEFINED),
-            (robocup_extension_pb2.OffensiveSide.SIDE_LEFT, Strategy.SIDE_LEFT),
-            (robocup_extension_pb2.OffensiveSide.SIDE_MIDDLE, Strategy.SIDE_MIDDLE),
-            (robocup_extension_pb2.OffensiveSide.SIDE_RIGHT, Strategy.SIDE_RIGHT),
-        )
+        self.set_state_defaults()
 
         self.tf_buffer = tf2_ros.Buffer()
         self.tf_listener = tf2_ros.TransformListener(self.tf_buffer, self.node)
 
-        # we will try multiple times till we manage to get a connection
-        while rclpy.ok() and self.socket is None:
-            self.socket = self.get_connection()
-            self.node.get_clock().sleep_for(Duration(seconds=1))
-            rclpy.spin_once(self.node)
-        self.socket.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
+        self.try_to_establish_connection()
+        self.run_spin_in_thread()
+
+        self.node.create_timer(1 / self.rate, self.send_message)
+        self.receive_forever()
 
+    def run_spin_in_thread(self):
         # Necessary in ROS2, else we are forever stuck receiving messages
-        thread = threading.Thread(target=rclpy.spin, args=(self.node), daemon=True)
+        thread = threading.Thread(target=rclpy.spin, args=[self.node], daemon=True)
```

Comment:
What do you mean by that? Is it ok this way?

Reply:
Due to this issue https://github.com/ros2/rclpy/issues/850 the clock and tf update callbacks are not isolated to separate threads. If they all use the same single threaded event loop, a TF transformation might get stuck while it is waiting for recent tf data, but the data cannot arrive because the waiting callback is blocking the event loop and the TF wait is not async. To make matters worse, it is also not timing out, because the clock is also not updated due to the blocked callback. A hacky way around this is the usage of a multithreaded executor. This runs multiple event loops in separate threads and is less likely to block as long as the thread count is high enough. 

The best fix, that nobody really wanted to do until now, is the implementation of https://github.com/ros2/rclcpp/pull/1218 for Python as described in https://github.com/ros2/rclpy/issues/850. Adding this would greatly improve the performance of our Python code (the Multi-Threaded Executor is kinda slow) and also clean up many deadlock issues.